app/src/main/java/com/tari/android/wallet/ui/fragment/send/AddRecipientFragment.kt

```diff
@@ -0,0 +1,517 @@
+        editText.clearFocus()
+    }
+
+    // region recycler view item touch listener
+
+    override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {
+        // no-op
+    }
+
+    /**
+     * Hide keyboard on list touch.
+     */
+    override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
+        val mActivity = activity ?: return false
+        UiUtil.hideKeyboard(mActivity)
+        editText.clearFocus()
+        return false
+    }
+
+    override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {
+        // no-op
+    }
+
+    // endregion
+
+    // region emoji id input masking
+
+    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
+        isDeletingSeparatorAtIndex =
+            if (count == 1 && after == 0 && s[start].toString() == emojiIdChunkSeparator) {
+                start
+            } else {
+                null
+            }
+    }
+
+    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+        // no-op
+    }
+
+    override fun afterTextChanged(editable: Editable) {
+        if (textWatcherIsRunning) {
+            return
+        }
+        textWatcherIsRunning = true
+
+        var text = editable.toString()
+        // if deleting a separator, first get the index of the character before that separator
+        // and delete that character
+        if (isDeletingSeparatorAtIndex != null) {
+            val index =
+                EmojiUtil.getStartIndexOfItemEndingAtIndex(text, isDeletingSeparatorAtIndex!!)
+            editable.delete(index, isDeletingSeparatorAtIndex!!)
+            isDeletingSeparatorAtIndex = null
+            text = editable.toString()
+        }
+        // delete all separators first
+        for ((offset, index) in EmojiUtil.getExistingChunkSeparatorIndices(
+            text,
+            emojiIdChunkSeparator
+        ).withIndex()) {
+            val target = index - (offset * emojiIdChunkSeparator.length)
+            editable.delete(target, target + 1)
+        }
+
+        continueButton.visibility = View.GONE
+        invalidEmojiIdTextView.visibility = View.GONE
+
+        val textWithoutSeparators = editable.toString()
+        if (textWithoutSeparators.firstNCharactersAreEmojis(emojiFormatterChunkSize)) {
+            editText.textAlignment = View.TEXT_ALIGNMENT_CENTER
+            editText.letterSpacing = inputEmojiIdLetterSpacing
+            // add separators
+            for ((offset, index) in EmojiUtil.getNewChunkSeparatorIndices(textWithoutSeparators).withIndex()) {
+                val target = index + (offset * emojiIdChunkSeparator.length)
+                editable.insert(target, emojiIdChunkSeparator)
+            }
+            // check if valid emoji - don't search if not
+            val numberofEmojis = textWithoutSeparators.numberOfEmojis()
+            if (textWithoutSeparators.containsNonEmoji() || numberofEmojis > emojiIdLength) {
```

Comment:
We should pull out a `validate_emoji_id(..)` method somewhere; the emoji id will also have a checksum character (it might already be in the FFI list??)  that will guard against fat finger errors etc

Reply:
I agree, I thought I would do it later when we get the spec. But I can pull out already, shall I?